src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp

```diff
@@ -545,12 +555,19 @@ static void copy_store_barrier(MacroAssembler* masm,
   }
 
   __ bind(done);
+}
 
-  if (new_ref != noreg) {
-    // Set store-good color, replacing whatever color was there before
-    __ ldr(tmp1, Address(rthread, ZThreadLocalData::store_good_mask_offset()));
-    __ bfi(new_ref, tmp1, 0, 16);
-  }
+static void copy_store_barrier_done(MacroAssembler* masm,
+                                    FloatRegister ref,
+                                    Address src,
+                                    Register tmp1,
+                                    Register tmp2) {
+  // Add good colors
+  __ orr(ref, Assembler::T16B, ref, z_copy_store_good_vreg);
+  // Store new values
+  __ umov(tmp1, ref, Assembler::D, 0);
+  __ umov(tmp2, ref, Assembler::D, 1);
+  __ stp(tmp1, tmp2, src);
```

Comment:
Why is there a store here? The barriers should not perform the actual stores. This seems inconsistent with the other code.

Reply:
I made the code look too much like x86...

x86 `ZBarrierSetAssembler::copy_store_at`:

```
if (done) {
  color source
=>  store source
} else if (fallback) {
  vector to general register
  copy_store_at for general register
}
```

aarch64:

Remove `Store new values` in the `ZBarrierSetAssembler::copy_store_at`: 

```
-
-  // Store new values
-  BarrierSetAssembler::copy_store_at(masm, decorators, type, bytes, dst, src1, src2, noreg, noreg, noreg, fnoreg, fnoreg, fnoreg);
```

But put the store into `copy_store_barrier_done`, then modified `ZBarrierSetAssembler::copy_store_at`:

```
if (is_dest_uninitialized) {
  copy_store_barrier_done
} else {
  copy_store_barrier
}
```

Because `if (done) else if (fallback)` is in the `copy_store_barrier` for aarch64, so I put a store here:

```
if (done) {
  copy_store_barrier_done
} else if (fallback) {
  vector to general
  copy_store_barrier for general
}
```

Reply:
Okay. It confuses me that there are now two functions called "copy_store_barrier_done", which suggests to me that the responsibility of said functions are similar. But they are completely different - one performs a store, and the other does not. It makes it difficult to read, and to convince myself that we are not storing to the same location twice.

Reply:
Sorry that I am not good at naming functions or my children. I will update my patch to make it better readability.